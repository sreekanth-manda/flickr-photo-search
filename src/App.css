body {
  background-color: black;
}

.App {
  text-align: center;
  display: flex;
}

.app-header {
  width: 100%;
  background-color: #942424;
  height: 120px;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
}

.app-title {
  font-size: 32px;
  color: white;
}

.app-content {
  display: flex;
  position: relative;
  padding: 15px;
  flex-direction: column;
  height: auto;
}

.prompt-search {
  font-size: 22px;
  font-weight: 600;
  align-self: center;
}


.app-loader {
  display: inline-block;
  width: 50px;
  height: 50px;
  border: 4px solid red;
  border-radius: 50%;
  border-top-color: rgb(12, 12, 12);
  animation: spin 1s ease-in-out infinite;
  -webkit-animation: spin 1s ease-in-out infinite;
  position: fixed;
  bottom: 0;
  left: 48%;
  margin: 10px;
}

@keyframes spin {
  to { -webkit-transform: rotate(360deg); }
}
@-webkit-keyframes spin {
  to { -webkit-transform: rotate(360deg); }
}