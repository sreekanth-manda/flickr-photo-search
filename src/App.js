import React from 'react';
import { throttle } from 'throttle-debounce';
import ImageList from './Components/ImageList/ImageList';
import SearchField from './Components/SearchField/SearchField';
import FullImage from './Components/FullImage/FullImage';
import flickrServices from './Services/FlickrService';
import { isScrollAreaAvailable, getImageUrl } from './Utils/Utils';
import "./App.css";

export default class App extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      searchText: '',
      imageList: [],
      pageNumber: 1,
      error: false,
      loading: false,
      showFullImage: false,
      currentImage: {}
    };

    this.onSearchInputChange = this.onSearchInputChange.bind(this);
    this.handleScroll = this.handleScroll.bind(this);
    this.onImageClick = this.onImageClick.bind(this);
    this.hideFullImage = this.hideFullImage.bind(this);
  }

  componentDidMount() {
    window.addEventListener('scroll', throttle(800, this.handleScroll));
  };

  componentWillUnmount() {
    window.removeEventListener('scroll', this.handleScroll);
  };

  handleScroll() {
    const promises = [];
    const images = [];

    if (!isScrollAreaAvailable()) {
      this.setState({ loading: true });

      flickrServices.searchFlickrImages(this.state.searchText, this.state.pageNumber + 1)
        .then(jsonResp => {
          jsonResp.body.photos.photo.forEach(photo => {
            promises.push(
              flickrServices.getPhoto(getImageUrl(photo))
                .then(() => images.push(photo))
                .catch((error) => console.log(error))
            );
          });

          Promise.all(promises).then(() => {
            this.setState((state, props) => ({
              pageNumber: jsonResp.body.photos.page,
              imageList: state.imageList.concat(images),
              error: false,
              loading: false
            }))
          });
        })
        .catch(err => this.setState({ error: true, loading: false }));
    }
  }

  onSearchInputChange(evt) {
    const searchText = evt.target.value;
    this.setState({ searchText, loading: true });

    flickrServices.searchFlickrImages(this.state.searchText)
      .then(jsonResp =>
        this.setState({
          imageList: jsonResp.body.photos.photo,
          error: false,
          loading: false
        }))
      .catch(err => this.setState({ error: true, loading: false }));
  }

  hideFullImage() {
    this.setState({ showFullImage: false });
  }

  onImageClick(imageData) {
    this.setState({ showFullImage: true, currentImage: imageData });
  }

  render() {
    return (
      <div className="app">
        <div className="app-header">
          <span className="app-title">Flickr Photo Search</span>
          <SearchField searchText={this.state.searchText} onSearchInputChange={this.onSearchInputChange} />
        </div>
        <div className="app-content" ref="appContent">
          {this.state.imageList.length !== 0 &&
            <ImageList
              images={this.state.imageList}
              onImageClick={this.onImageClick}
            />
          }
          {this.state.loading && <div className='app-loader' />}
        </div>
        {this.state.showFullImage && <FullImage hideFullImage={this.hideFullImage} image={this.state.currentImage} />}
      </div>
    );
  }
}
