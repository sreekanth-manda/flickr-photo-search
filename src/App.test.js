import React from 'react';
import { mount } from 'enzyme';
import App from './App';
import flickrServices from './Services/FlickrService';

const photos = {
  body: {
    photos: {
      photo: [
        { id: 'blah1' },
        { id: 'blah2' }
      ]
    }
  }
};

describe('App', () => {
  it('renders without crashing', () => {
    const wrapper = mount(<App />);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render SearchField component', () => {
    const wrapper = mount(<App />);
    expect(wrapper.find('SearchField').length).toBe(1);
  });

  it('should set state with the input search text value ', () => {
    const wrapper = mount(<App />);
    wrapper.find('.search-input').simulate('change', { target: { value: 'Hello' } });
    expect(wrapper.state().searchText).toBe('Hello');
  });

  it('should set state with the images list from the flickr service ', async () => {
    const wrapper = mount(<App />);
    flickrServices.searchFlickrImages = jest.fn().mockResolvedValue(photos);
    await wrapper.find('.search-input').simulate('change', { target: { value: 'Hello' } });
    expect(wrapper.state().imageList).toMatchObject([...photos.body.photos.photo]);
  });
});
