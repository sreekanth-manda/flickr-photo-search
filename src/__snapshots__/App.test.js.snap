// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders without crashing 1`] = `"<div class=\\"app\\"><div class=\\"app-header\\"><span class=\\"app-title\\">Flickr Photo Search</span><input type=\\"text\\" class=\\"search-input\\" placeholder=\\"Search images here\\" value=\\"\\"></div><div class=\\"app-content\\"></div></div>"`;
